src/reason_pprint_ast.ml

```diff
@@ -482,133 +527,239 @@ let rec sequentialIfBlocks x =
    let x = tail & head;
  *)
 
+(* "Almost Simple Prefix" function applications parse with the rule:
+
+   `PREFIXOP simple_expr %prec below_DOT_AND_SHARP`, which in turn is almost
+   considered a "simple expression" (it's acceptable anywhere a simple
+   expression is except in a couple of edge cases.
+
+   "Unary Prefix" function applications parse with the rule:
+
+   `MINUS epxr %prec prec_unary_minus`, which in turn is considered an
+   "expression" (not simple). All unary operators are mapped into an identifier
+   beginning with "~".
+
+   TODO: Migrate all "almost simple prefix" to "unsary prefix". When `!`
+   becomes "not", then it will make more sense that !myFunc (arg) is parsed as
+   !(myFunc arg) instead of (!myFunc) arg.
 
-let prefix_symbols  = [ '!'; '?'; '~' ] ;;
+ *)
+let almost_simple_prefix_symbols  = [ '!'; '?'; '~'] ;;
+(* Subset of prefix symbols that have special "unary precedence" *)
+let unary_minus_prefix_symbols  = [ "~-"; "~-."] ;;
+let unary_plus_prefix_symbols  = ["~+"; "~+." ] ;;
 let infix_symbols = [ '='; '<'; '>'; '@'; '^'; '|'; '&'; '+'; '-'; '*'; '/';
                       '$'; '%'; '\\' ]
+(* There are just a couple of infix operators that start with `!`. All others that
+   start with ! would be prefix, so we keep a list of exceptions here. *)
+let infix_starting_with_bang = ["!="; "!=="]
```

Comment:
I check it later on to make sure it's not interpreted as prefix - but I think I can just check special_infix_strings in that place. I'll try to remove this.